app/org/sagebionetworks/bridge/dynamodb/DynamoTopicSubscriptionDao.java

```diff
@@ -0,0 +1,116 @@
+        hashKey.setRegistrationGuid(registration.getGuid());
+        
+        DynamoDBQueryExpression<DynamoTopicSubscription> query = new DynamoDBQueryExpression<DynamoTopicSubscription>()
+                .withConsistentRead(false).withHashKeyValues(hashKey);
+
+        return mapper.queryPage(DynamoTopicSubscription.class, query).getResults();
+    }
+    
+    public TopicSubscription subscribe(NotificationRegistration registration, NotificationTopic topic) {
+        checkNotNull(registration);
+        checkNotNull(topic);
+
+        // If it already exists, this will just overwrite it.
+        DynamoTopicSubscription subscription = new DynamoTopicSubscription();
+        subscription.setRegistrationGuid(registration.getGuid());
+        subscription.setTopicGuid(topic.getGuid());
+        
+        SubscribeRequest request = new SubscribeRequest()
+                .withEndpoint(registration.getEndpointARN())
+                .withProtocol("application")
+                .withTopicArn(topic.getTopicARN());
+        
+        SubscribeResult result = snsClient.subscribe(request);
+        String subscriptionARN = result.getSubscriptionArn();
+        subscription.setSubscriptionARN(subscriptionARN);
+        
+        // The integrity issue we want to avoid is where SNS has registered the user to receive
+        // a topic, but we have no DDB record. We cannot unsubscribe them at that point until 
+        // they re-subscribe, then try again.
+        try {
+            
+            mapper.save(subscription);
+            return subscription;
+            
+        } catch(Throwable throwable) {
+            snsClient.unsubscribe(subscriptionARN);
+            throw throwable;
+        }
+    }
+
+    public void unsubscribe(NotificationRegistration registration, NotificationTopic topic) {
+        checkNotNull(registration);
+        checkNotNull(topic);
+
+        DynamoTopicSubscription hashKey = new DynamoTopicSubscription();
+        hashKey.setRegistrationGuid(registration.getGuid());
+        hashKey.setTopicGuid(topic.getGuid());
+        
+        DynamoTopicSubscription subscription = mapper.load(hashKey);
+        if (subscription == null) {
+            throw new EntityNotFoundException(TopicSubscription.class);
+        }
+        
+        snsClient.unsubscribe(subscription.getSubscriptionARN());
+        
+        // If this fails, user is told there was an error, and there is an orphaned DDB record. They can try and 
+        // unsubscribe again and that will work. We also attempt to reconcile the records in the service.
+        mapper.delete(subscription);
+    }
+
+    @Override
+    public void delete(TopicSubscription subscription) {
+        checkNotNull(subscription);
+        // No exception is caught here, so user will see errors, but nevertheless we will
+        // try and clean up both data stores.
+        try {
+            snsClient.unsubscribe(subscription.getSubscriptionARN());
```

Comment:
The only place this is called is if we catch a NotFoundException when loading SNS topics. So we kind of expect that it will fail again and what we want is to get rid of the mapping record.

Nothing in this method indicates this how it is being used... I'll try and rename for clarity.